Add build instructions for cranelift backend as part of Rust repo (#111827)
rust-lang/rust: compiler/rustc_codegen_cranelift/Readme.md

For additional ways to use rustc_codegen_cranelift like the JIT mode see [usage.md](docs/usage.md).

## Building and testing with changes in rustc code

This is useful when changing code in `rustc_codegen_cranelift` as part of changing [main Rust repository](https://github.com/rust-lang/rust/).
This can happen, for example, when you are implementing a new compiler intrinsic.

Instruction below uses `$RustCheckoutDir` as substitute for any folder where you cloned Rust repository.

> You need to do this steps to successfully compile and use cranelift backend with your changes in rustc code:

Reviewer: ```suggestion
You need to do this steps to successfully compile and use the cranelift backend with your changes in rustc code:
```

1. `cd $RustCheckoutDir`
2. Run `python x.py setup` and choose option for compiler (`b`).

> 3. Build compiler and necessary tools: `python x.py build --stage=2 compiler library/std src/tools/rustdoc src/tools/cargo src/tools/rustfmt`

Reviewer: ```suggestion
3. Build compiler and necessary tools: `python x.py build --stage=2 compiler library/std src/tools/rustdoc src/tools/rustfmt`
```
You can skip compiling cargo by copying the cargo of another toolchain. It is a bit unfortunate that building rustfmt is necessary, but I don't know of any other way to full assemble the stage2 sysroot.

> 4. Copy exectutable files from `./build/host/stage2-tools/<your hostname triple>/release`

Reviewer: ```suggestion
4. Copy cargo into the stage2 toolchain: `cp $(rustup which cargo) .build/<your hostname triple>/stage2/bin/cargo`
```

to `./build/host/stage2/bin/`. Note that you would need to do this every time you rebuilt `rust` repository.
5. Link your new `rustc` to toolchain: `rustup toolchain link stage2 ./build/host/stage2/`.

> 6. (Optional) compile y.rs: `rustc +stage2 -O y.rs`.

Reviewer: ```suggestion
6. (Windows only) compile y.rs: `rustc +stage2 -O y.rs` and use `./y` instead of `./y.rs`.
```

> 7. You need to prefix every `./y` command by `rustup run stage2` to make cranelift use your local changes in rustc.

Reviewer: ```suggestion
7. You need to prefix every `./y.rs` command by `rustup run stage2` to make cg_clif use your local changes in rustc.
```

>   * `rustup run stage2 ./y prepare`

Reviewer: ```suggestion
  * `rustup run stage2 ./y.rs prepare`
```

>   * `rustup run stage2 ./y build --sysroot cliff`

Reviewer: ```suggestion
  * `rustup run stage2 ./y.rs build`
```

`--sysroot clif` is the default.

>   * (Optional) run tests: `rustup run stage2 ./y test --sysroot clif`

Reviewer: ```suggestion
  * (Optional) run tests: `rustup run stage2 ./y.rs test`
```

> 8. Now you can use your cranelift build to compile other Rust programs, e.g. you can open any Rust crate and run commands like `$RustCheckoutDir/compiler/rustc_codegen_cranelift/dist/cargo-clif.exe build --release`.

Reviewer: ```suggestion
8. Now you can use your cg_clif build to compile other Rust programs, e.g. you can open any Rust crate and run commands like `$RustCheckoutDir/compiler/rustc_codegen_cranelift/dist/cargo-clif build --release`.
```
